pennylane/workflow/construct_batch.py

```diff
@@ -73,6 +74,102 @@ def _get_full_transform_program(
+    num_user_transforms: int,
+) -> slice:
+    """Interpret the level specification for the initial user transform slice.
+
+    This function handles slicing into the user transforms only, before any
+    gradient or device transforms are applied.
+
+    Args:
+        level: The level specification from user input
+        num_user_transforms: Number of user transforms
+
+    Returns:
+        slice: The slice to apply to the user transform program
+    """
+    if level == "top":
+        return slice(0, 0)
+
+    if level in ("user", "device", "gradient"):
+        return slice(0, num_user_transforms)
+
+    if isinstance(level, str):  # pragma: no cover
+        raise ValueError(
+            f"level {level} not recognized. Acceptable strings are 'device', 'top', 'user', and 'gradient'."
+        )
+
+    if level is None:
+        return slice(0, num_user_transforms)
+
+    if isinstance(level, int):
+        return slice(0, min(level, num_user_transforms))
+
+    # Handle slice objects - clamp to user transform bounds
+    if isinstance(level, slice):
+        start = level.start if level.start is not None else 0
+        # stop = level.stop if level.stop is not None else num_user_transforms
+        # if stop >= 0:
+        #     stop = min(stop, num_user_transforms)
+        return slice(start, level.stop, level.step)
+
+    return level  # pragma: no cover
+
+
+# pylint: disable=too-many-return-statements
+def _interpret_level_inner(
+    level: Union[Literal["top", "user", "device", "gradient"], int, slice, None],
+    num_user_transforms: int,
+    has_gradient_expand: bool,
+) -> slice:
+    """Interpret the level specification for the inner transform slice.
+
+    This function handles slicing into the remaining transforms (gradient + device)
+    after user transforms have already been applied.
+
+    Args:
+        level: The level specification from user input
+        num_user_transforms: Number of user transforms (already applied)
+        has_gradient_expand: Whether gradient expansion transform exists
+
+    Returns:
+        slice: The slice to apply to the remaining transform program
+    """
+    start = max(0, num_user_transforms)
+    if level in ("top", "user"):
+        return slice(start, 0)  # No additional transforms needed
+
+    if level == "gradient":
+        end_idx = int(has_gradient_expand)
+        return slice(start, num_user_transforms + end_idx)
+
+    if level == "device":
+        return slice(start, None)  # Include all remaining transforms
+
+    if isinstance(level, str):  # pragma: no cover
+        raise ValueError(
+            f"level {level} not recognized. Acceptable strings are 'device', 'top', 'user', and 'gradient'."
+        )
+
+    if level is None:
+        return slice(start, None)  # Include all remaining transforms
+
+    if isinstance(level, int):
+        # Calculate how many transforms beyond user transforms are needed
+        return slice(start, level)
```

Comment:
```suggestion
    if level is None or isinstance(level, int):
        return slice(start, level)  # Include all remaining transforms
```
Might as well combine.